src/stores/cpauth/middleware.ts

```diff
@@ -0,0 +1,39 @@
+import CPAuth from 'lib/CPAuth/CPAuth';
+import OAuth2UserImpl from 'lib/OAuth2/OAuth2User';
+import { AnyAction, Middleware } from 'redux';
+import { loadUserError, userExpired } from 'stores/cpauth/actions';
+import { CPAUTH_USER_EXPIRED, CPAUTH_USER_LOAD } from 'stores/cpauth/constants';
+import { getCPAuthUser } from 'stores/cpauth/selectors';
+
+export function cpAuthMiddleware(cpAuth: CPAuth): Middleware {
```

Comment:
This middleware is used to try renew user information, if a user is expired.

In general, the user will be renewed automatically, but the flow will fail if there's a race condition between renewal and a separate request. This prevents that.

This will also prevent actions from getting dispatched, if the user expired.